### PR TITLE
Added code to handle multi-graph in mst

### PR DESCRIPTION
**Pull request by Aditya-Shandilya1182**

### CODE REVIEW

**Comment by Aditya-Shandilya1182** [comment]
This code is written to address issue #7451.

**Comment by Aditya-Shandilya1182** [comment]
Code used to test the feature:
````python
for g in nx.SpanningTreeIterator(nx.cycle_graph(3,create_using=nx.MultiGraph)):
    print(g.edges())
````
Result before changes:
fli


Result after changes:
mst

**Comment by Aditya-Shandilya1182** [comment]
I will try to add the test cases.

File: networkx/algorithms/tree/mst.py [hunk 1]
```diff
@@ -1125,19 +1125,34 @@ def _write_partition(self, partition):
...
         Removes partition data from the graph
         """
-        for u, v, d in G.edges(data=True):
-            if self.partition_key in d:
-                del d[self.partition_key]
+        partition_key = self.partition_key
+        edges = (
+            G.edges(keys=True, data=True) if G.is_multigraph() else G.edges(data=True)
+        )
+        for *e, d in edges:
```
Comment (eriknw) [inline networkx/algorithms/tree/mst.py:1153]: I like this pattern! What do you think about using it above when setting `d[partition_key]`? This way, we wouldn't need separate for loops for `if G.is_multigraph()` and `else`.

File: networkx/algorithms/tree/mst.py [hunk 2]
```diff
@@ -1125,19 +1125,34 @@ def _write_partition(self, partition):
...
+        if G.is_multigraph():
+            for u, v, k, d in G.edges(keys=True, data=True):
+                edge = (u, v, k)
+                d[partition_key] = partition_dict.get(edge, EdgePartition.OPEN)
+
+        # Iterate over edges for simple graphs
+        else:
+            for u, v, d in G.edges(data=True):
+                edge = (u, v)
+                d[partition_key] = partition_dict.get(edge, EdgePartition.OPEN)
```
Comment (Peiffap) [inline networkx/algorithms/tree/mst.py:1143]: With Erik's suggestion.
```suggestion
        edges = G.edges(keys=True, data=True) if G.is_multigraph() else G.edges(data=True)
        for *e, d in edges:
            d[partition_key] = partition_dict.get(e, EdgePartition.OPEN)
```
I don't think this should affect performance.

**Review by Peiffap — COMMENTED** [review]
Thanks! I like how you've refactored things cleanly. Small suggestion below based on Erik's review.

It'd be nice to have some test cases for this new behavior too. Could you add those?

This isn't new to this PR, but could you also fix the typo Dan mentioned (it should say "undirected" here)?
https://github.com/networkx/networkx/blob/4dca07982f5b0c9fee852e828ac1f0186d296ed7/networkx/algorithms/tree/mst.py#L1015-L1016

File: networkx/algorithms/tree/mst.py [hunk 3]
```diff
@@ -1125,19 +1125,34 @@ def _write_partition(self, partition):
...
-            if (u, v) in partition.partition_dict:
-                d[self.partition_key] = partition.partition_dict[(u, v)]
-            else:
-                d[self.partition_key] = EdgePartition.OPEN
+
+        partition_dict = partition.partition_dict
+        partition_key = self.partition_key
+        G = self.G
+
+        # Iterate over edges with for multi-graphs
```
Comment (Peiffap) [inline networkx/algorithms/tree/mst.py:1133]:
```suggestion
```
Oops, I didn't catch this line in my other suggestion.

File: networkx/algorithms/tree/mst.py [hunk 4]
```diff
@@ -1125,19 +1125,26 @@ def _write_partition(self, partition):
...
-            else:
-                d[self.partition_key] = EdgePartition.OPEN
+
+        partition_dict = partition.partition_dict
+        partition_key = self.partition_key
+        G = self.G
+
+        edges = G.edges(keys=True, data=True) if G.is_multigraph() else G.edges(data=True)
+        for *e, d in edges:
+            d[partition_key] = partition_dict.get(e, EdgePartition.OPEN)
```
Comment (Peiffap) [inline networkx/algorithms/tree/mst.py:1135]: Whoops, that's my bad.

```suggestion
        edges = G.edges(keys=True, data=True) if G.is_multigraph() else G.edges(data=True)
        for *e, d in edges:
            d[partition_key] = partition_dict.get(tuple(e), EdgePartition.OPEN)
```

File: networkx/algorithms/tree/mst.py [hunk 5]
```diff
@@ -1125,19 +1125,26 @@ def _write_partition(self, partition):
...
         """
         Removes partition data from the graph
         """
-        for u, v, d in G.edges(data=True):
-            if self.partition_key in d:
-                del d[self.partition_key]
+        partition_key = self.partition_key
+        edges = (
+            G.edges(keys=True, data=True) if G.is_multigraph() else G.edges(data=True)
+        )
```
Comment (Peiffap) [inline networkx/algorithms/tree/mst.py:1144]: While we're here, I think this looks cleaner. Feel free to disregard.

```suggestion
        edges = G.edges(keys=True, data=True) if G.is_multigraph() else G.edges(data=True)
```

**Review by Peiffap — COMMENTED** [review]
This should fix the test failures.

**Review by mjschwenne — APPROVED** [review]

**Comment by Peiffap** [comment]
Erik's suggestion was fine, I just didn't think it through/test it locally when I implemented it! Thanks for taking a look at this!

**Review by dschult — COMMENTED** [review]
This looks good!  Thanks @Aditya-Shandilya1182 @mjschwenne and @Peiffap !

Could we revert the switch in the test file for `nx.is_tree` -> `is_tree`?  The idiom `nx.is_tree` leaves no doubt where it comes from, and simplifies/stabilizes the import at the top of the file. (The long chain of subpackages is fragile to changes in our codebase organization. It is also not strictly needed, but then again, neither is the import statement as `nx` is already defined.)

More succinctly, could we switch `is_tree` back to `nx.is_tree` and remove the import statement?  Thanks...

**Pull request merged**